### PR TITLE
Fix various problems with importing modules

### PR DESCRIPTION
**Pull request by adamretter**
This back-ports some fixes to eXist-db from FusionDB around the handling of `import module`.

**NOTE**: The PR https://github.com/eXist-db/exist/pull/3579 needs to be merged **BEFORE** this one!

1.  Closes https://github.com/eXist-db/exist/issues/1174
2. Now raises correct errors for all cases of:
    1. XQST0047
    2. XQST0034
    3. XQST0049

### CODE REVIEW

File: exist-core/src/main/java/org/exist/source/AbstractSource.java [hunk 1]
```diff
@@ -26,6 +26,8 @@
 import java.io.InputStreamReader;
 import java.nio.charset.Charset;
 
+import net.jpountz.xxhash.XXHash64;
```
Comment (dizzzz) [inline exist-core/src/main/java/org/exist/source/AbstractSource.java:29]: is this a duplicate from another PR?

File: exist-core/src/main/java/org/exist/xquery/FunctionFactory.java [hunk 2]
```diff
@@ -263,48 +267,77 @@ private static JavaCall javaFunctionBinding(XQueryContext context,
...
-                fn = getInternalModuleFunction(context, ast, params, qname, module);
-            } else {
-                //Function is from an imported XQuery module
-                fn = getXQueryModuleFunction(context, ast, params, qname, module);
+        final Module[] modules = context.getModules(uri);
+        if (modules != null) {
+            // Function might belongs to a module
+            for (int i = 0; i < modules.length; i++) {
+                final Module module = modules[i];
+                final boolean throwOnNotFound = i == modules.length - 1;
```
Comment (dizzzz) [inline exist-core/src/main/java/org/exist/xquery/FunctionFactory.java:279]: for me :-) unreadable line

File: exist-core/src/main/java/org/exist/xquery/functions/util/ModuleInfo.java [hunk 3]
```diff
@@ -109,7 +111,7 @@
 			new QName("get-module-description", UtilModule.NAMESPACE_URI, UtilModule.PREFIX),
 			"Returns a short description of the module identified by the namespace URI.",
 			new SequenceType[] { NAMESPACE_URI_PARAMETER },
-			new FunctionReturnSequenceType(Type.STRING, Cardinality.EXACTLY_ONE, "the description of the active function module identified by the namespace URI"),
+			new FunctionReturnSequenceType(Type.STRING, Cardinality.ONE_OR_MORE, "the description of the active function module identified by the namespace URI"),
```
Comment (dizzzz) [inline exist-core/src/main/java/org/exist/xquery/functions/util/ModuleInfo.java:114]: need to check: update of monex needed/possible?

**Review by dizzzz — APPROVED** [review]
LGTM ; I hope the 2 PRs add-up...

**Comment by dizzzz** [comment]
small issue found
image

File: exist-core/src/main/java/org/exist/source/AbstractSource.java (hunk 1 above)
Comment (adamretter) [reply exist-core/src/main/java/org/exist/source/AbstractSource.java:29]: This PR is atop #3579

**Comment by adamretter** [comment]
@dizzzz Ooof! Looks like I forgot to backport one of the fixes from FusionDB... I have now added it to this PR :-)

**Comment by dizzzz** [comment]
in some builds...

image

**Review by dizzzz — CHANGES_REQUESTED** [review]
not all builds are ok :-(

**Comment by dizzzz** [comment]
For Java11 and 14 the same issue with "Module":

image

**Comment by dizzzz** [comment]
@adamretter I cannot apply this 
[patch.diff.txt](https://github.com/eXist-db/exist/files/5447548/patch.diff.txt)

**Comment by adamretter** [comment]
@dizzzz Should be fixed now. JDK 9+ has its own `Module` class...

**Comment by dizzzz** [comment]
yes I figured out about java9...

**Review by dizzzz — APPROVED** [review]

**Comment by dizzzz** [comment]
Where is the change? looks not in this PR?

**Comment by adamretter** [comment]
@dizzzz It is the commit titled "[bugfix] Fix compilation on JDK 9+"

**Comment by dizzzz** [comment]
right :-) I was looking for the import change, but this is indeed better

**Comment by dizzzz** [comment]
Wohooooo

**Pull request merged**